File: Mobile/lib/data/s3_assets.dart
/// S3 image URLs shared with Web/Admin (`shared/s3Assets.json`).
class S3Assets {
  S3Assets._();

  static const String publicBaseUrl =
      'https://rezervame-assets-abs.s3.ap-southeast-2.amazonaws.com';

  static const String bucket = 'rezervame-assets-abs';

  static const Map<String, String> defaultCategories = {
    'hairService':
        'https://rezervame-assets-abs.s3.ap-southeast-2.amazonaws.com/uploads/defaults/categories/hairService.jpg',
    'spaService':
        'https://rezervame-assets-abs.s3.ap-southeast-2.amazonaws.com/uploads/defaults/categories/spaService.jpg',
    'nailCare':
        'https://rezervame-assets-abs.s3.ap-southeast-2.amazonaws.com/uploads/defaults/categories/nailCare.jpg',
    'beautyService':
        'https://rezervame-assets-abs.s3.ap-southeast-2.amazonaws.com/uploads/defaults/categories/beautyService.jpg',
    'barber':
        'https://rezervame-assets-abs.s3.ap-southeast-2.amazonaws.com/uploads/defaults/categories/barber.jpg',
    'Massage':
        'https://rezervame-assets-abs.s3.ap-southeast-2.amazonaws.com/uploads/defaults/categories/Massage.jpg',
    'massage':
        'https://rezervame-assets-abs.s3.ap-southeast-2.amazonaws.com/uploads/defaults/categories/massage.jpg',
    'tattoo':
        'https://images.unsplash.com/photo-1560707303-4e980ce876ad?q=80&w=600&fit=crop',
    'yoga':
        'https://images.unsplash.com/photo-1544367567-0f2fcb009e0b?q=80&w=600&fit=crop',
    'estetica':
        'https://images.unsplash.com/photo-1516975080664-ed2fc6a32937?q=80&w=600&fit=crop',
    'dermatology':
        'https://images.unsplash.com/photo-1576091160399-112ba8d25d1d?q=80&w=600&fit=crop',
  };

  static String? defaultCategoryUrl(String key) {
    final k = key.trim();
    if (k.isEmpty) return null;
    return defaultCategories[k] ?? defaultCategories[k.toLowerCase()];
  }

  static bool isS3PublicUrl(String? url) {
    if (url == null || url.trim().isEmpty) return false;
    return url.startsWith(publicBaseUrl) || url.contains('$bucket.s3.');
  }
}

File: Web/src/lib/s3Assets.ts
import s3Assets from "../../../shared/s3Assets.json";

export const S3_PUBLIC_BASE_URL = s3Assets.publicBaseUrl;

export const DEFAULT_CATEGORY_IMAGES: Record<string, string> = {
  ...s3Assets.defaultCategories,
};

/** Legacy hostname from early S3 setup docs — objects live on the current bucket with the same key. */
const LEGACY_S3_HOSTS = [
  "rezervame-assets-yourname.s3.ap-southeast-2.amazonaws.com",
  "rezervame-assets-yourname.s3.amazonaws.com",
];

export function normalizePublicImageUrl(url: string | null | undefined): string {
  const raw = (url ?? "").trim();
  if (!raw) return "";
  let s = raw;
  let correctHost: string;
  try {
    correctHost = new URL(S3_PUBLIC_BASE_URL.replace(/\/+$/, "")).host;
  } catch {
    correctHost = `${s3Assets.bucket}.s3.ap-southeast-2.amazonaws.com`;
  }
  for (const legacy of LEGACY_S3_HOSTS) {
    if (s.includes(legacy)) s = s.split(legacy).join(correctHost);
  }
  if (s.includes("rezervame-assets-yourname")) {
    s = s.replace(/rezervame-assets-yourname/g, s3Assets.bucket);
  }
  // S3 keys are case-sensitive; lowercase massage.jpg was never uploaded.
  if (s.includes("/defaults/categories/massage.jpg")) {
    s = s.replace("/defaults/categories/massage.jpg", "/defaults/categories/Massage.jpg");
  }
  return s;
}

export function isS3PublicUrl(url: string | null | undefined): boolean {
  if (!url?.trim()) return false;
  const t = normalizePublicImageUrl(url);
  return t.startsWith(S3_PUBLIC_BASE_URL) || t.includes(`${s3Assets.bucket}.s3.`);
}

export function defaultCategoryImageForKey(key: string): string {
  const k = key.trim();
  return DEFAULT_CATEGORY_IMAGES[k] ?? DEFAULT_CATEGORY_IMAGES[k.toLowerCase()] ?? "";
}
